### PR TITLE
Added optnet for reducing memory consumption

### PR DESCRIPTION
**Pull request by melgor**
### What does this PR do?
Reduce memory consumption thanks to https://github.com/fmassa/optimize-net. For example, memory consumption from nn4 model was reduced by 33%

### How should this PR be tested?
It should be tested by running several training procedure, using different models and batch size

### What are the relevant issues?
#106 

# Questions:
@fmassa currently you method works slightly worse than pure "shareGradient" idea. Do you know why? (difference is ~50 Mb in model which use 3.5 Gb)
@bamos should this method be run as default or chosen by command line option. I was thinking about only because for using it, the user need to install additional dependence

+ Does this PR add new (Python) dependencies?

This method use https://github.com/fmassa/optimize-net and it should be installed.

### CODE REVIEW

File: training/util.lua [hunk 1]
```diff
@@ -54,3 +54,14 @@ function receiveTensor(obj, buffer)
    end
    return buffer
 end
+
+--Reduce the memory consumption by model by sharing the buffers
+function optimizeNet( model, inputSize )
+   optnet = require 'optnet'
```
Comment (fmassa) [inline training/util.lua]: just a heads up, those variables can be made local without problem.

**Comment by fmassa** [comment]
@melgor I made some small improvements on the optimization part in https://github.com/fmassa/optimize-net/pull/5, it gives a few percent improvement over current version (and should close the gap with the hand-assigned `shareGradient` from fb.resnet.torch. I just need to update the README with new values and I'll merge the PR.
I'll have a closer look at this integration tomorrow. I'm thinking about using it in inference mode as well (when appropriate), etc.

**Comment by fmassa** [comment]
@melgor regarding dependencies, the only dependency is on `graph`, which comes with standard `torch` distributions,

**Comment by fmassa** [comment]
What about using a `pcall` in `require 'optnet'` and skipping the memory optimization path if the user doesn't have `optnet` installed in their machine ?
This will avoid initial breaks due to missing dependency (which will probably happen when the users update their `openface` code and their code won't work anymore with `module 'optnet' not found` errors). What do you think ?

**Comment by bamos** [comment]
@fmassa - I like that idea. @melgor can you make this print a warning if optnet isn't installed? Then the old CI tests should pass. I'll then re-build the Docker container with `optnet`.

**Comment by melgor** [comment]
I use the 'pcall' for checking if package is installed. 
Additionally, I add brief message about not having this package, as it is very usefull

**Comment by bamos** [comment]
Thanks @melgor - looks great!

**Comment by fmassa** [comment]
Cool ! Thanks @melgor !

**Pull request merged**